subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClassLoaderUtils.java

```diff
@@ -216,24 +201,36 @@ public Package[] getPackages(ClassLoader classLoader) {
             return GET_PACKAGES_METHOD.invoke(classLoader);
         }
 
+        @Nullable
         @Override
         public Package getPackage(ClassLoader classLoader, String name) {
             return GET_PACKAGE_METHOD.invoke(classLoader, name);
         }
     }
 
-    private static class LookupPackagesFetcher extends AbstractClassLoaderLookuper implements ClassLoaderPackagesFetcher {
-        private MethodType getPackagesMethodType = MethodType.methodType(Package[].class, new Class<?>[]{});
-        private MethodType getDefinedPackageMethodType = MethodType.methodType(Package.class, new Class<?>[]{String.class});
+    private static class Java9PackagesFetcher implements ClassLoaderPackagesFetcher {
 
         @Override
         public Package[] getPackages(ClassLoader classLoader) {
-            return invoke(classLoader, "getPackages", getPackagesMethodType);
+            return new ClassLoader(classLoader) {
```

Comment:
Do you need to change this to make the code work again without `--add-opens`? Is the equivalent public API calling `classLoader.getDefinedPackages()`?  If so, we can just call that here because this class is only used for Java 9+

Reply:
This code does not get hit in the test workers apparently -- at least in the simple case. It is used in the `MultiParentClassLoader` to implement `getPackage` and `getPackages`. Who knows whether or not that code is actually exercised. Though, if it were to get exercised without these changes, the previous code would be performing deep reflection on jdk internals, so we definitely need a way to avoid that.

One problem (or rather, difference) with `getDefinedPackages` is that it does not consider the parent class loaders at all. Really, it seems we need to make further updates to `MultiParentClassLoader` and override `getDefinedPackages` there instead of overriding `getPackages` -- that way we can use just the public API and won't have to deal with anything `protected`. Though, this might get tricky with different java versions (we're already dealing with this in `ClassLoaderUtils`), I think this is probably a better solution. I'll take a close look at this and see if we can avoid the needless extending of `ClassLoader` just to call the protected method.

Reply:
TBH I don't really think there is a more elegant way of doing this. I plan to leave this as-is unless you've got other ideas.